File: testdata/jennies/builders/struct_fields_as_args_assignment/TypescriptBuilder/src/sandbox/someStructBuilder.gen.ts
import * as cog from '../cog';
import * as sandbox from '../sandbox';

export class SomeStructBuilder implements cog.Builder<sandbox.SomeStruct> {
    protected readonly internal: sandbox.SomeStruct;

    constructor() {
        this.internal = sandbox.defaultSomeStruct();
    }

    /**
     * Builds the object.
     */
    build(): sandbox.SomeStruct {
        return this.internal;
    }

    time(from: string,to: string): this {
        if (!this.internal.time) {
            this.internal.time = {
	from: "now-6h",
	to: "now",
};
        }
        this.internal.time.from = from;
        this.internal.time.to = to;
        return this;
    }
}

File: testdata/jennies/builders/struct_fields_as_args_assignment/GoBuilder/sandbox/somestruct_builder_gen.go
package sandbox

import (
	cog "github.com/grafana/cog/generated/cog"
)

var _ cog.Builder[SomeStruct] = (*SomeStructBuilder)(nil)

type SomeStructBuilder struct {
    internal *SomeStruct
    errors cog.BuildErrors
}

func NewSomeStructBuilder() *SomeStructBuilder {
	resource := NewSomeStruct()
	builder := &SomeStructBuilder{
		internal: resource,
		errors: make(cog.BuildErrors, 0),
	}

	return builder
}



func (builder *SomeStructBuilder) Build() (SomeStruct, error) {
	if err := builder.internal.Validate(); err != nil {
		return SomeStruct{}, err
	}
	
	if len(builder.errors) > 0 {
	    return SomeStruct{}, cog.MakeBuildErrors("sandbox.someStruct", builder.errors)
	}

	return *builder.internal, nil
}

func (builder *SomeStructBuilder) Time(from string,to string) *SomeStructBuilder {
if builder.internal.Time == nil {
    builder.internal.Time = &struct {
    From string `json:"from"`
    To string `json:"to"`
}{}
}
    builder.internal.Time.From = from
    builder.internal.Time.To = to

    return builder
}

